# Wu-2023-soilsampling

## Overview
Aggregative boot cover sampling may be a more representative, practical, and powerful method for preharvest produce soil testing than grab sampling because boot covers aggregate soil from larger areas. Our study tests if boot cover sampling results reflect quality and safety indicator organisms and community diversity of grab sampling. We collected soil samples from commercial romaine lettuce fields spanning 5060 m2 using boot covers (n = 28, m = 1.1 ± 0.4 g; wearing boot covers and walking along the path), composite grabs (n = 28, m = 231 ± 24 g; consisting of 60 grabs of 3–5 g each), and high-resolution grabs (n = 72, m = 56 ± 4 g; taking one sample per stratum). Means and standard deviations of log-transformed aerobic plate counts (APCs) were 7.0 ± 0.3, 7.1 ± 0.2, and 7.3 ± 0.2 log(CFU/g) for boot covers, composite grabs, and high-resolution grabs, respectively. APCs did not show biologically meaningful differences between sample types. Boot covers recovered on average 0.6 log(CFU/g) more total coliforms than both grabs (p < 0.001) where means and standard deviations of log-transformed counts were 3.2 ± 1.0, 2.6 ± 0.6, and 2.6 ± 1.0 log(CFU/g) for boot covers, composite grabs, and high-resolution grabs, respectively. There were no generic _E. coli_ detected in any sample by enumeration methods with LODs of 1.3–2.1 log(CFU/g) for boot covers and 0.5 log(CFU/g) for both grabs. By 16S rRNA sequencing, community species diversity (alpha diversity) was not significantly different within collection methods. While communities differed (p < 0.001) between soil sampling methods (beta diversity), variance in microbial communities was not significantly different. Of the 28 phyla and 297 genera detected, 25 phyla (89%) and 258 genera (87%) were found by all methods. Overall, aggregative boot cover sampling is similar to both grab methods for recovering quality and safety indicator organisms and representative microbiomes. This justifies future work testing aggregative soil sampling for foodborne pathogen detection.

## Usage
### Setup
Mention the environment the code was run on during development and testing as well as any dependencies that are needed.

### Running
Explain in detailed steps how to run the code in order to reproduce the results shown above in the results section.

## Authors
You can view the list of authors in the [AUTHORS](/AUTHORS) file.

## Contact
Corresponding author: Matthew J. Stasiewicz<br>
103 Agricultural Bioprocess Lab<br>
1302 W. Pennsylvania<br>
Urbana, IL, 1361801<br>
USA<br>
+1-217-265-0963<br>
[mstasie@illinois.edu](mailto:mstasie@illinois.edu)

## Citation
Wu, J., R. J. Gathman, J. Quintanilla Portillo, C. Gaulke, M. Kim, and M. J. Stasiewicz. 2023. Aggregative Soil Sampling Using Boot Covers Compared to Soil Grabs From Commercial Romaine Fields Shows Similar Indicator Organism and Microbial Community Recoveries. Journal of Food Protection. 86:100177.

## License
This project's code is licensed under the GNU General Public License v3.0 and dataset is licensed the Creative Commons Attribution Share Alike 4.0 International license. Please see the [LICENSE.code](/LICENSE.code) and [LICENSE.dataset](/LICENSE.dataset) files for details.

## Acknowledgements
Include acknowledgements relevant to the project here.

## Funding
This study was funded by the USDA National Institute of Food and Agriculture, Agricultural and Food Research Initiative Competitive Program – Food Safety and Defense Program, grant number: 
2021-67017-33828 https://portal.nifa.usda.gov/web/crisprojectpages/1024000-aggregative-sampling-for-powerful-preharvest-leafy-green-food-safety-testing.html.
